lib/EarlyFunctionAnalysis/DetectABIMFI.h

```diff
@@ -0,0 +1,212 @@
+  using SetOfRegisters = std::set<model::Register::Values>;
+  using SUL = SetUnionLattice<SetOfRegisters>;
+
+  SetOfRegisters ArgumentRegisters;
+  SetOfRegisters ReturnRegisters;
+
+  [[nodiscard]] FunctionABI combineValues(const FunctionABI &A2) const {
+    FunctionABI Result = {
+      .ArgumentRegisters = SUL::combineValues(ArgumentRegisters,
+                                              A2.ArgumentRegisters),
+      .ReturnRegisters = SUL::combineValues(ReturnRegisters, A2.ReturnRegisters)
+    };
+
+    return Result;
+  }
+
+  [[nodiscard]] bool isLessOrEqual(const FunctionABI &A2) const {
+    if (!SUL::isLessOrEqual(ArgumentRegisters, A2.ArgumentRegisters)) {
+      return false;
+    }
+
+    return SUL::isLessOrEqual(ReturnRegisters, A2.ReturnRegisters);
+  }
+};
+
+class LatticeElement {
+private:
+  using map_t = std::map<MetaAddress, FunctionABI>;
+  using iterator_t = map_t::iterator;
+  using const_iterator_t = map_t::const_iterator;
+
+  map_t Map;
+
+public:
+  const_iterator_t begin() const { return Map.begin(); }
+  iterator_t begin() { return Map.begin(); }
+
+  const_iterator_t end() const { return Map.end(); }
+  iterator_t end() { return Map.end(); }
+
+  iterator_t find(const map_t::key_type &Key) { return Map.find(Key); }
+  const_iterator_t find(const map_t::key_type &Key) const {
+    return Map.find(Key);
+  }
+
+  map_t::mapped_type &operator[](const auto &Key) {
+    return Map[Key];
+  }
+
+  map_t::mapped_type &at(const map_t::key_type &Key) {
+    return Map.at(Key);
+  }
+
+  const map_t::mapped_type &at(const map_t::key_type &Key) const {
+    return Map.at(Key);
+  }
+};
+
+class DetectABIMFI {
+private:
+  GeneratedCodeBasicInfo &GCBI;
+  std::map<MetaAddress, OutlinedFunction> &OutlinedFunctions;
+  CFGAnalyzer &Analyzer;
+  TupleTree<model::Binary> &Binary;
+
+public:
+  using Label = BasicBlockNode *;
+
+  explicit DetectABIMFI(
+    GeneratedCodeBasicInfo &GCBI,
+    std::map<MetaAddress, OutlinedFunction> &OutlinedFunctions,
+    CFGAnalyzer &Analyzer,
+    TupleTree<model::Binary> &Binary) :
+    GCBI{ GCBI },
+    OutlinedFunctions{ OutlinedFunctions },
+    Analyzer{ Analyzer },
+    Binary{ Binary } {}
+
+  [[nodiscard]] LatticeElement
+  combineValues(const LatticeElement &E1, const LatticeElement &E2) const
+  {
+    LatticeElement Result;
+
+    for (const auto &[Address, E1Abi] : E1) {
+      auto E2It = E2.find(Address);
+      if (E2It != E2.end()) {
+        Result[Address] = E1Abi.combineValues(E2It->second);
+      } else {
+        Result[Address] = E1Abi;
+      }
+    }
+
+    for (const auto &[Address, E2Abi] : E2) {
+      auto ResultIt = Result.find(Address);
+      if (ResultIt == Result.end()) {
+        Result[Address] = E2Abi;
+      }
+    }
+
+    return Result;
+  }
+
+  [[nodiscard]] bool
+  isLessOrEqual(const LatticeElement &E1, const LatticeElement &E2) const {
+    for (const auto &[Address, E1Abi] : E1) {
+      auto It = E2.find(Address);
+      if (It != E2.end()) {
```

Comment:
Are you sure it's OK to ignore the situation in which `Address` is not in `E2`?
What should the default value be in case of absence?